ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/AppendOnlyCommandDeduplicationParallelIT.scala

```diff
@@ -16,76 +16,145 @@ import com.daml.ledger.api.testtool.infrastructure.Assertions.fail
+      val okResponses = responses.getOrElse(Code.OK, 0)
+      val alreadyExistsResponses = responses.getOrElse(Code.ALREADY_EXISTS, 0)
+      // Participant-based command de-duplication can currently also reject duplicates via a SQL exception when using the CommandSubmissionService
+      val internalResponses = responses.getOrElse(Code.INTERNAL, 0)
+      // Canton can return ABORTED for duplicate submissions
+      // Participant based command de-duplication can currently also return ABORTED when using the CommandService
+      val abortedResponses = responses.getOrElse(Code.ABORTED, 0)
+      val duplicateResponses =
+        alreadyExistsResponses + internalResponses + abortedResponses
+      assert(
+        okResponses == 1 && duplicateResponses == numberOfParallelRequests - 1,
+        s"Expected $expectedDuplicateResponses duplicate responses and one accepted, got $responses",
+      )
+      assert(activeContracts.size == 1)
+    }
+  }
+
+  private def submitAndWaitRequestAndGetStatusCode(
+      ledger: ParticipantTestContext
+  )(request: SubmitAndWaitRequest, parties: Party*)(implicit ec: ExecutionContext) = {
+    val submissionId = UUID.randomUUID().toString
+    val requestWithSubmissionId = request.update(_.commands.submissionId := submissionId)
+    val submitResult = ledger.submitAndWait(requestWithSubmissionId)
+    submissionResultToFinalStatusCode(ledger)(submitResult, submissionId, parties: _*)
+  }
   protected def submitRequestAndGetStatusCode(
       ledger: ParticipantTestContext
   )(request: SubmitRequest, parties: Party*)(implicit ec: ExecutionContext): Future[Code] = {
     val submissionId = UUID.randomUUID().toString
     val requestWithSubmissionId = request.update(_.commands.submissionId := submissionId)
-    ledger
+    val submitResult = ledger
       .submit(requestWithSubmissionId)
-      .flatMap(_ => ledger.findCompletion(parties: _*)(_.submissionId == submissionId))
-      .map {
-        case Some(completion) =>
-          completion.getStatus.code
-        case None => fail(s"Did not find completion for request with submission id $submissionId")
-      }
-      .recover {
-        case GrpcException(status, _) =>
-          status.getCode.value()
-        case otherException => fail("Not a GRPC exception", otherException)
+    submissionResultToFinalStatusCode(ledger)(submitResult, submissionId, parties: _*)
+  }
+
+  private def submissionResultToFinalStatusCode(
+      ledger: ParticipantTestContext
+  )(submitResult: Future[Unit], submissionId: String, parties: Primitive.Party*)(implicit
+      ec: ExecutionContext
+  ) = {
+    submitResult
+      .transformWith {
+        case Failure(exception) =>
+          exception match {
+            case GrpcException(status, _) =>
+              Future.successful(status.getCode)
+            case NonFatal(otherException) =>
+              fail(s"Not a GRPC exception $otherException", otherException)
```

Comment:
Why include the exception twice?

```suggestion
              fail(s"Not a GRPC exception.", otherException)
```